parser-typechecker/src/Unison/Codebase/MainTerm.hs

```diff
@@ -43,23 +44,30 @@ getMainTerm loadTypeOfTerm parseNames mainName mainType =
       let refs = NamesWithHistory.lookupHQTerm hq (NamesWithHistory.NamesWithHistory parseNames mempty)
       let a = Parser.Ann.External
       case toList refs of
+        [] -> pure (NotFound mainName)
         [Referent.Ref ref] -> do
           typ <- loadTypeOfTerm ref
           case typ of
             Just typ ->
-              if Typechecker.isSubtype typ mainType
+              if Typechecker.fitsScheme typ mainType
                 then do
                   let tm = DD.forceTerm a a (Term.ref a ref)
                   return (Success hq tm typ)
                 else pure (BadType mainName $ Just typ)
             _ -> pure (BadType mainName Nothing)
-        _ -> pure (NotFound mainName)
+        _ -> pure (error "multiple matching refs") -- TODO: make a real exception
 
--- '{io2.IO, Exception} ()
+-- forall x. '{ io2.IO, Exception } x
 builtinMain :: Var v => a -> Type.Type v a
-builtinMain a = Type.arrow a (Type.ref a DD.unitRef) io
+builtinMain a =
+  let x = Var.named "x"
+   in Type.forall a x (builtinMain' a (Type.var a x))
+
+-- '{io2.IO, Exception} res
+builtinMain' :: Var v => a -> Type.Type v a -> Type.Type v a
```

Comment:
How does this differ from `builtinMain`?

AFAICT it accepts an explicit result type, if that's the case I'd suggest `builtinMainWithResultType`;

Or failing that, a description in the docs would be 👌🏼  😄